### ISSUE TITLE
FYI: Emacs 26 will remove frame-local variables

### ISSUE THREAD

**Issue by npostavs** [nex3/perspective-el#64]
I believe this will break perspective-el.

https://lists.gnu.org/archive/html/emacs-devel/2016-12/msg00843.html

**Comment by Gastove** [comment]
Just hit this after updating to a new nightly of Emacs 26; definitely breaks perspective.el.

**Comment by nex3** [comment]
It should be possible to work around this by creating a map from frames to a struct that contains everything that's currently a frame-local variable. I'm not sure when I'll have time to do that, though.

**Comment by redshirtrob** [comment]
It's less than ideal, but if you're not actually using multiple frames, you can add this:
```elisp
(when (not (fboundp 'make-variable-frame-local))
  (defun make-variable-frame-local (variable) variable))
```

**Comment by john2x** [comment]
I did a bit of work in removing frame-local variables from the codebase in my fork at https://github.com/john2x/perspective-el.
I work with multiple frames extensively (via [nameframe](https://github.com/john2x/nameframe) package), so I needed perspective to be future proof :)

Initial tests seems good so far on Arch Linux, Emacs 25.3. Had a bit of trouble around the global-mode-string before I found out about `:eval` form. I'll do a bit more testing before opening a pull request.

**Comment by nex3** [comment]
> I'll do a bit more testing before opening a pull request.

That would be great, thanks!

**Comment by arichiardi** [comment]
Trying the @john2x branch as well (thanks!) and reporting back.

**Comment by arichiardi** [comment]
@nex3 Don't know if this is the right place but with your fork I get an error:

```
Debugger entered--Lisp error: (void-variable perspectives-hash)
  (gethash name perspectives-hash)
  (let* ((name (file-name-nondirectory (directory-file-name project-to-switch))) (persp (gethash name perspectives-hash))) (cond ((and persp (not (equal persp persp-curr))) (persp-switch name)) ((not persp) (let ((frame (selected-frame))) (persp-switch name) (projectile-switch-project-by-name project-to-switch) (if (not (equal frame (selected-frame))) (progn (let ((old-frame (selected-frame)) (old-buffer (current-buffer))) (unwind-protect (progn (select-frame frame 'norecord) (persp-kill name)) (if (frame-live-p old-frame) (progn (select-frame old-frame 'norecord))) (if (buffer-live-p old-buffer) (progn (set-buffer old-buffer)))))))))))
  projectile-persp-switch-project("~/git/project/")
  funcall-interactively(projectile-persp-switch-project "~/git/project/")
  call-interactively(projectile-persp-switch-project nil nil)
  command-execute(projectile-persp-switch-project)
```

It is probably linked to [`projectile-persp`](https://github.com/bbatsov/persp-projectile).

**Comment by john2x** [comment]
@arichiardi yes, I think projectile-persp will need to update its usage of `perspectives-hash` to something like `(frame-parameter nil 'perspectives-hash)`

**Comment by arichiardi** [comment]
Opened a PR in there: https://github.com/bbatsov/persp-projectile/pull/15

Everything works fine with your branch and my branch on my machine.

**Comment by thisch** [comment]
Fixed in #67. Can be closed now
